compiler/rustc_query_system/src/dep_graph/graph.rs

```diff
@@ -535,6 +533,32 @@ impl<D: Deps> DepGraph<D> {
         }
     }
 
+    /// This encodes a diagnostic by creating a node with an unique index and assoicating
+    /// `diagnostic` with it, for use in the next session.
+    #[inline]
+    pub fn record_diagnostic<Qcx: QueryContext>(&self, qcx: Qcx, diagnostic: &DiagInner) {
+        if let Some(ref data) = self.data {
+            D::read_deps(|task_deps| match task_deps {
+                TaskDepsRef::EvalAlways | TaskDepsRef::Ignore => return,
+                TaskDepsRef::Forbid | TaskDepsRef::Allow(..) => {
```

Comment:
Shouldn't Forbid be unreachable here? If we are decoding a query will we ever emit diagnostics?

Reply:
The forbid case will just panic in `read_index`, so it's just a sanity check.

Reply:
I guess 😃